okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt

```diff
@@ -163,29 +200,90 @@ class HttpLoggingInterceptor
 
     @Throws(IOException::class)
     override fun intercept(chain: Interceptor.Chain): Response {
-      val level = this.level
+      var request = chain.request()
+
+      val level = this.requestLevel(request)
 
-      val request = chain.request()
       if (level == Level.NONE) {
         return chain.proceed(request)
       }
 
-      val logBody = level == Level.BODY
-      val logHeaders = logBody || level == Level.HEADERS
+      val logHeaders = level == Level.BODY || level == Level.HEADERS
 
-      val requestBody = request.body
+      val decompressionAlgorithms = findCompressionAlgorithms(chain)
 
       val connection = chain.connection()
+
+      val requestBody = request.body
       var requestStartMessage =
         ("--> ${request.method} ${redactUrl(request.url)}${if (connection != null) " " + connection.protocol() else ""}")
-      if (!logHeaders && requestBody != null) {
+      if (!logHeaders && level != Level.STREAMING && requestBody != null) {
         requestStartMessage += " (${requestBody.contentLength()}-byte body)"
       }
       logger.log(requestStartMessage)
 
-      if (logHeaders) {
-        val headers = request.headers
+      if (level != Level.BASIC) {
+        request = logRequest(request, level, decompressionAlgorithms)
+      }
+
+      val startNs = System.nanoTime()
+      val response: Response
+      try {
+        response = chain.proceed(request)
+      } catch (e: Exception) {
+        val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
+        logger.log(
+          buildString {
+            append("<-- HTTP FAILED: $e.")
+            append(" ${redactUrl(request.url)} (${tookMs}ms)")
+          },
+        )
+        throw e
+      }
+
+      val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
+
+      val responseBody = response.body
+      val contentLength = responseBody.contentLength()
+      val bodySize = if (contentLength != -1L) "$contentLength-byte" else "unknown-length"
+      logger.log(
+        buildString {
+          append("<-- ${response.code}")
+          if (response.message.isNotEmpty()) append(" ${response.message}")
+          append(" ${redactUrl(response.request.url)} (${tookMs}ms")
+          if (!logHeaders) append(", $bodySize body")
+          append(")")
+        },
+      )
 
+      return if (level != Level.BASIC) {
+        logResponse(
+          response,
+          decompressionAlgorithms,
+          startNs,
+          contentLength,
+        )
+      } else {
+        response
+      }
+    }
+
+    private fun findCompressionAlgorithms(chain: Interceptor.Chain): List<DecompressionAlgorithm> {
+      val compressionInterceptor =
+        (chain.call() as? RealCall)?.client?.interceptors?.find { it is CompressionInterceptor } as? CompressionInterceptor
```

Comment:
I am uncomfortable with this . . . having the logging interceptor take over decompression is super confusing. We can get plaintext logging by applying this one earlier than that one in the pipeline?